### PR TITLE
fix(ec2): improve errors around subnet selection

### LINKED ISSUE #3859: AutoScalingGroup vpc_subnets doesn't seem to do anything
### What is the problem?
I'm using cdk with Python and have an AutoScalingGroup launching an EC2 instance within a vpc. Everything was working fine, except when I tried to add `vpc_subnets` to the constructor in order to specify which subnet I want the instance to launch in, `cdk synth` resulted in no differences to the CloudFormation templates and `cdk diff` doesn't reflect that either. 

### Reproduction Steps

If you have code like the following, the `vpc_subnets` parameter doesn't seem to be honoured:

    aws_autoscaling.AutoScalingGroup(
        self, "MyAutoscalingGroup", 
        vpc=my_vpc,
        instance_type=aws_ec2.InstanceType.of(aws_ec2.InstanceClass.STANDARD5, aws_ec2.InstanceSize.XLARGE),
        machine_image=aws_ec2.WindowsImage(
            version=aws_ec2.WindowsVersion.WINDOWS_SERVER_2019
        ),
        vpc_subnets=aws_ec2.SubnetSelection(subnet_name="subnet-0af00000000000000"),  # (using valid subnet ID here)
    )

### Environment

  - **CDK CLI Version:** 1.6.0
  - **Module Version:** aws-cdk.aws-autoscaling==1.6.0
  - **OS:** Ubuntu
  - **Language:** Python

### PR DESCRIPTION
**Pull request by rix0rrr**
Selecting subnets by VPC and type is a confusing topic with some
API misapprehensions.

Try to fix that by renaming `subnetName` to `subnetGroupName`, and
giving more clear error messages about legal values in case of misuse.

Closes #3859.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-ec2/lib/vpc.ts [hunk 1]
```diff
@@ -162,22 +162,40 @@ export enum SubnetType {
  */
 export interface SubnetSelection {
   /**
-   * Place the instances in the subnets of the given type
+   * Select all subnets of the given type
    *
-   * At most one of `subnetType` and `subnetName` can be supplied.
+   * At most one of `subnetType` and `subnetGroupName` can be supplied.
```
Comment (eladb) [inline packages/@aws-cdk/aws-ec2/lib/vpc.ts]: Should this be a union-like instead?

File: packages/@aws-cdk/aws-ec2/lib/vpc.ts [hunk 2]
```diff
@@ -162,22 +162,40 @@ export enum SubnetType {
...
 
   /**
-   * Place the instances in the subnets with the given name
+   * Select the subnet group with the given name
    *
-   * (This is the name supplied in subnetConfiguration).
+   * Select the subnet group with the given name. This only needs
+   * to be used if you have multiple subnet groups of the same type
+   * and you need to distinguish between them. Otherwise, prefer
+   * selection by type.
```
Comment (eladb) [inline packages/@aws-cdk/aws-ec2/lib/vpc.ts]: "prefer selection by type" => "prefer `subnetType`"

File: packages/@aws-cdk/aws-ec2/lib/vpc.ts [hunk 3]
```diff
@@ -1407,30 +1444,24 @@ class ImportedVpc extends VpcBase {
  * Returns "private subnets" by default.
  */
 function reifySelectionDefaults(placement: SubnetSelection): SubnetSelection {
-  if (placement.subnetType !== undefined && placement.subnetName !== undefined) {
+  if (placement.subnetName !== undefined) {
+    if (placement.subnetGroupName !== undefined) {
+      throw new Error(`Cannot supply both 'subnetName' and 'subnetGroupName', they mean the same thing. Prefer 'subnetGroupName'.`);
```
Comment (eladb) [inline packages/@aws-cdk/aws-ec2/lib/vpc.ts]: It should read `Please use only "subnetGroupName" ("subnetName" is deprecated and has the same behavior)`

File: packages/@aws-cdk/aws-ec2/lib/vpc.ts [hunk 4]
```diff
@@ -1407,30 +1444,24 @@ class ImportedVpc extends VpcBase {
...
+    }
+    placement = {...placement, subnetGroupName: placement.subnetName };
+  }
+
+  if (placement.subnetType !== undefined && placement.subnetGroupName !== undefined) {
     throw new Error('Only one of subnetType and subnetName can be supplied');
```
Comment (eladb) [inline packages/@aws-cdk/aws-ec2/lib/vpc.ts]: `subnetName` => `subnetGroupName`

File: packages/@aws-cdk/aws-ec2/lib/vpc.ts (hunk 1 above)
Comment (rix0rrr) [reply packages/@aws-cdk/aws-ec2/lib/vpc.ts]: Maybe. Bit hard to change right now, and then the right thing to make union-like is the whole `subnetSelection` object. I predict there will be many more subnet selection requirements in the future.

**Pull request merged**